go/teams/chain.go

```diff
@@ -1602,9 +1645,35 @@ func (t *teamSigchainPlayer) addInnerLink(
 			return res, fmt.Errorf("unsupported link type: %s", payload.Body.Type)
 		}
 	}
-	if isHighLink {
-		res.newState.inner.LastHighLinkID = link.LinkID().Export()
-		res.newState.inner.LastHighSeqno = link.Seqno()
+
+	// We only run addInnerLink out of order when inflating, and so we skip
+	// computing high skips during it.
+	// updates high skip data, so we skip it. If highSkip is presented and
+	// either we don't need to reverify or we want to check it, check it.
+	highSkip := payload.HighSkip
+	if highSkipVerifyDesired {
+		if highSkip != nil && !reverifyNeeded {
+			if highSkip.Seqno != res.newState.GetLatestHighSeqno() {
+				return res, fmt.Errorf("Expected HighSkipSeqno %d, got %d @ %d", res.newState.inner.LastHighSeqno, highSkip.Seqno, payload.Seqno)
+			}
+			lastHighLinkID := res.newState.GetLatestHighLinkID()
+			if string(lastHighLinkID) == "" {
+				if highSkip.Hash != nil {
+					return res, fmt.Errorf("Expected nil HighSkipHash, got %s", *highSkip.Hash)
+				}
+			} else {
+				if highSkip.Hash == nil {
+					return res, fmt.Errorf("Expected non-nil HighSkipHash, got nil")
+				}
+				if *highSkip.Hash != string(lastHighLinkID) {
+					return res, fmt.Errorf("Expected HighSkipHash=%s, got %s", string(lastHighLinkID), *highSkip.Hash)
+				}
+			}
+		}
+		if isHighLink {
+			res.newState.inner.LastHighLinkID = link.LinkID().Export()
+			res.newState.inner.LastHighSeqno = link.Seqno()
+		}
```

Comment:
You might pull this green section into its own function to avoid adding to the behemoth? I don't care though.